controllers/storagecluster/cephnfs.go

```diff
@@ -0,0 +1,340 @@
+		return reconcile.Result{}, nil
+	}
+
+	cephNFS, err := r.newCephNFSInstance(instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	existingCephNFS := cephv1.CephNFS{}
+	err = r.Client.Get(context.TODO(), types.NamespacedName{Name: cephNFS.Name, Namespace: cephNFS.Namespace}, &existingCephNFS)
+	switch {
+	case err == nil:
+		if existingCephNFS.DeletionTimestamp != nil {
+			r.Log.Info("Unable to restore CephNFS because it is marked for deletion.", "CephNFS", klog.KRef(existingCephNFS.Namespace, existingCephNFS.Name))
+			return reconcile.Result{}, fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existingCephNFS.Name)
+		}
+
+		r.Log.Info("Restoring original CephNFS.", "CephNFS", klog.KRef(cephNFS.Namespace, cephNFS.Name))
+		existingCephNFS.ObjectMeta.OwnerReferences = cephNFS.ObjectMeta.OwnerReferences
+		existingCephNFS.Spec = cephNFS.Spec
+		err = r.Client.Update(context.TODO(), &existingCephNFS)
+		if err != nil {
+			r.Log.Error(err, "Unable to update CephNFS.", "CephNFS", klog.KRef(cephNFS.Namespace, cephNFS.Name))
+			return reconcile.Result{}, err
+		}
+	case errors.IsNotFound(err):
+		r.Log.Info("Creating CephNFS.", "CephNFS", klog.KRef(cephNFS.Namespace, cephNFS.Name))
+		err = r.Client.Create(context.TODO(), cephNFS)
+		if err != nil {
+			r.Log.Error(err, "Unable to create CephNFS.", "CephNFS", klog.KRef(cephNFS.Namespace, cephNFS.Name))
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// ensureDeleted deletes the CephNFS resource owned by the StorageCluster
+func (obj *ocsCephNFS) ensureDeleted(r *StorageClusterReconciler, sc *ocsv1.StorageCluster) (reconcile.Result, error) {
+	foundCephNFS := &cephv1.CephNFS{}
+	cephNFS, err := r.newCephNFSInstance(sc)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	err = r.Client.Get(context.TODO(), types.NamespacedName{Name: cephNFS.Name, Namespace: sc.Namespace}, foundCephNFS)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("Uninstall: CephNFS not found.", "CephNFS", klog.KRef(cephNFS.Namespace, cephNFS.Name))
+			return reconcile.Result{}, nil
+		}
+		r.Log.Error(err, "Uninstall: Unable to retrieve CephNFS.", "CephNFS", klog.KRef(cephNFS.Namespace, cephNFS.Name))
+		return reconcile.Result{}, fmt.Errorf("uninstall: Unable to retrieve CephNFS %v: %v", cephNFS.Name, err)
+	}
+
+	if cephNFS.GetDeletionTimestamp().IsZero() {
+		r.Log.Info("Uninstall: Deleting CephNFS.", "CephNFS", klog.KRef(foundCephNFS.Namespace, foundCephNFS.Name))
+		err = r.Client.Delete(context.TODO(), foundCephNFS)
+		if err != nil {
+			r.Log.Error(err, "Uninstall: Failed to delete CephNFS.", "CephNFS", klog.KRef(foundCephNFS.Namespace, foundCephNFS.Name))
+			return reconcile.Result{}, fmt.Errorf("uninstall: Failed to delete CephNFS %v: %v", foundCephNFS.Name, err)
+		}
+	}
+
+	err = r.Client.Get(context.TODO(), types.NamespacedName{Name: cephNFS.Name, Namespace: sc.Namespace}, foundCephNFS)
```

Comment:
why one more get again here?

Reply:
> why one more get again here?

This get is to make sure, the resource is deleted. 
If the resource is still present, waiting error is returned to trigger backoff reconcile, please see line 120-121 below.

Reply:
isnt Delete a blocking call?

Reply:
> isnt Delete a blocking call?

I am not sure, I'll get back after testing it. 

Meanwhile, Other types' `ensureDeleted` follow the same format of checking the resource's existence after deletion.

Reply:
No, the delete is not blocking.

```
{"level":"info","ts":1650279363.4743314,"logger":"controllers.StorageCluster","msg":"Uninstall: Deleting CephBlockPool.","Request.Namespace":"openshift-storage","Request.Name":"ocs-storagecluster","CephBlockPool":"openshift-storage/ocs-storagecluster-cephblockpool"}
{"level":"error","ts":1650279363.4865177,"logger":"controllers.StorageCluster","msg":"Uninstall: Waiting for CephBlockPool to be deleted.","Request.Namespace":"openshift-storage","Request.Name":"ocs-storagecluster","CephBlockPool":"openshift-storage/ocs-storagecluster-cephblockpool","stacktrace":"github.com/red-hat-storage/ocs-operator/controllers/storagecluster.(*StorageClusterReconciler).deleteResources\n\t/remote-source/app/controllers/storagecluster/uninstall_reconciler.go:333\ngithub.com/red-hat-storage/ocs-operator/controllers/storagecluster.(*StorageClusterReconciler).reconcilePhases\n\t/remote-source/app/controllers/storagecluster/reconcile.go:334\ngithub.com/red-hat-storage/ocs-operator/controllers/storagecluster.(*StorageClusterReconciler).Reconcile\n\t/remote-source/app/controllers/storagecluster/reconcile.go:161\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/remote-source/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:298\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/remote-source/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:253\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/remote-source/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:214"}
```

The r.Log.Error just before last return statement in every `ensureDelete` func is called with a `nil` error which leads to a panic.